### PR TITLE
Adds the ability to run bash script through appium-for-mac

### PR DESCRIPTION
**Pull request by davidmokos**
we can now run bash script on the target machine running the appium-for-mac using the selenium driver.execute_script("mybashcommand")

### CODE REVIEW

File: AppiumForMac/Server/Handlers/AfMHandlers.m [hunk 1]
```diff
@@ -240,6 +240,19 @@ - (AppiumMacHTTPJSONResponse *)post_url:(NSString*)path data:(NSData*)postData
 // POST /session/:sessionId/execute
 // Inject a snippet of JavaScript into the page for execution in the context of the currently selected frame. The executed script is assumed to be synchronous and the result of evaluating the script is returned to the client.
 
+- (AppiumMacHTTPJSONResponse *)post_execute:(NSString*)path data:(NSData*)postData
+{
+    return [self executeWebDriverCommandWithPath:path data:postData onMainThread:YES commandBlock:^(AfMSessionController *session, NSDictionary *commandParams, int *statusCode)
+            {
+                // The bash command to run
+                NSString *command = (NSString*)[commandParams objectForKey:@"script"];
```
Comment (stuartbrussell-intuit) [inline AppiumForMac/Server/Handlers/AfMHandlers.m]: Rename to 'script' or 'scriptString', or something else besides 'command'. In this context, command refers to the web driver command itself.

File: AppiumForMac/Server/Handlers/AfMHandlers.m [hunk 2]
```diff
@@ -240,6 +240,19 @@ - (AppiumMacHTTPJSONResponse *)post_url:(NSString*)path data:(NSData*)postData
...
+
```
Comment (stuartbrussell-intuit) [inline AppiumForMac/Server/Handlers/AfMHandlers.m]: If you can, please support the full "execute" command by allowing shell script arguments passed in through the 'args' commandParam. That will allow the same scriptString to have different behaviors without having to rebuild the string in your client script.

**Comment by mykola-mokhnach** [comment]
this might be a perfect backdoor %/

**Comment by stuartbrussell-intuit** [comment]
@mykola-mokhnach Yeah, AfM opens an unprotected port. In that sense AfM has been a form of back door since day one. This just improves that "feature".   😄

**Comment by mykola-mokhnach** [comment]
Do we want to keep it like this? I'd rather vote for closing such hole rather than extending it.

**Comment by stuartbrussell-intuit** [comment]
Would that mean a password-protected https port? I thought WebDriver ports were by definition open. Could be wrong. But if an open port is a bad thing, then AfM is a bad thing. AfM is unique in the sense that the driver's universe is the whole machine, rather than the inside of a web view. Does AfM therefore not meet a particular security standard? I'm all for making it more secure if we can, but a faulty password scheme is almost worse than none, because it makes people think it is "safe".

Curious: does Appium proper deal with this? Does it give you access to the whole device, or just one app at a time?

**Comment by stuartbrussell-intuit** [comment]
Another possibility is to set a user default on the AfM machine, so that only a specific set of IP addresses can connect.

**Comment by mykola-mokhnach** [comment]
@stuartbrussell-intuit In all the other driver we do limit the access to the command line, so the user cannot execute arbitrary shell commands on the target machine. Am I right @jlipps ?

**Comment by davidmokos** [comment]
I committed some changes, bat I need to dig into the API to find a replacement for the NSTask launch method, which is deprecated but there is apparently no substitution.

About the security concern - what blocks me from opening the Terminal app and pass a command into it by driving the UI?

**Comment by jlipps** [comment]
Yeah I think it's less about secure access to AfM and more about what users can do once they have a session. With other drivers we try to limit what can be done via the driver to something other than general-purpose computing on the host machine :-)

The question I usually ask when people want this kind of feature is what their use case is. Often we can then simply build a feature specifically for that use case, and not blow open a huge security hole.

**Comment by stuartbrussell-intuit** [comment]
It's fine if we reject this change, but... this goes to the heart of AfM. Because of the way it is implemented, AfM has access to the entire UI of the machine, not just a specific app. So in its current form, it could easily launch Terminal and type into the window using standard webdriver commands. That's actually valuable because it can control multi-app work flows, but with great power comes great... security holes? 😄    Seriously though, how should we treat AfM knowing how powerful it is. (btw, was this a surprise to people? I was explicit in my SelConf 2017 presentation that AfM is a fundamentally different beast from (e.g.) ChromeDriver, but not everyone saw it.)

**Comment by jlipps** [comment]
I guess if it's well-known that AfM can control the entire computer anyway then this doesn't add any security hole! So I'm fine with it in that case.

**Comment by stuartbrussell-intuit** [comment]
👍  However, I do think it would be prudent to add something that disallows access from any other machine, for people who want to restrict who can control the computer. (e.g. a white list).

https://github.com/appium/appium-for-mac/issues/32

**Comment by stuartbrussell-intuit** [comment]
@davidmokos I'm totally swamped, give me a couple of days and then I can review your latest. If the code is good, I'll approve it.

**Comment by mykola-mokhnach** [comment]
I think we could use this feature together with the flag from https://github.com/appium/appium/pull/9790, which should be implemented first on osx driver side

**Comment by stuartbrussell-intuit** [comment]
@mykola-mokhnach Just to clarify for my confused brain: The fix for #9790 would mean, before "unsafe" actions,  the user has to set a preference on the server side, or launch the server from the command line with an arg? So by default it would not be allowed? Just making sure I understand.

**Comment by mykola-mokhnach** [comment]
@stuartbrussell-intuit Yes, your understanding is correct. The idea is that remote clients don't have access to the remote server, so they cannot change command line arguments there, but it should not be a problem if one runs Appium in a local environment. Check, for example, https://github.com/appium/appium-android-driver/pull/293

File: AppiumForMac/Server/Handlers/AfMHandlers.m [hunk 3]
```diff
@@ -240,6 +240,17 @@ - (AppiumMacHTTPJSONResponse *)post_url:(NSString*)path data:(NSData*)postData
 // POST /session/:sessionId/execute
 // Inject a snippet of JavaScript into the page for execution in the context of the currently selected frame. The executed script is assumed to be synchronous and the result of evaluating the script is returned to the client.
 
+- (AppiumMacHTTPJSONResponse *)post_execute:(NSString*)path data:(NSData*)postData
```
Comment (mykola-mokhnach) [inline AppiumForMac/Server/Handlers/AfMHandlers.m]: it should be also possible to pass additional environment variables

File: AppiumForMac/Server/Controller/AfMSessionController.m [hunk 4]
```diff
@@ -574,6 +574,32 @@ -(void) closeWindow
     [self.currentWindow performAction:@"AXCancel"];
 }
 
+- (NSString *)executeShellScript:(NSString *)script
+{
+    NSTask *task = [[NSTask alloc] init];
```
Comment (mykola-mokhnach) [inline AppiumForMac/Server/Controller/AfMSessionController.m]: I assume NSTask starts with an empty environment by default. Shouldn't we inherit the current system env by default?

**Comment by Jiurong-Yang** [comment]
Is anyone working on this? Just to clarify the CRs for this pull request are:
1). Replace the NSTask module with something that is not deprecated. I assume a popen module should suffice? 
2). Comply to the execute mobile command API structure
http://appium.io/docs/en/commands/mobile-command/
3). Comply with the --relaxed-security server flag. 

Not sure what is the current state of this project. I'd like some feedback if possible. Thank you
@mykola-mokhnach 
@stuartbrussell-intuit

**Comment by mykola-mokhnach** [comment]
Hi @Jiurong-Yang . Thanks for your feedback
Unfortunately we don't have enough time and resources to support this driver properly, as we are mostly focused on mobile stuff. We'd be happy if there were more contributors to it.

**Comment by mykola-mokhnach** [comment]
Superseded by https://github.com/appium/appium-mac-driver/pull/38

**Pull request closed without merging**